src/git.md

```diff
@@ -0,0 +1,181 @@
+rust-lang/rust that existed the last time you ran `git pull rust master` on
+your master branch. As such, when you submit your PR it is possible that some
+of the changes that have been made to rust-lang/rust since then are in conflict
+with the changes you've made; maybe someone else changed the same lines of
+code, or git cannot figure out how to merge your changes with the others for
+another reason.
+
+When this happens, you need to resolve the conflicts before your changes can be
+merged. First, get a local copy of the conflicting changes. Checkout your local
+master branch with `git checkout master`. Then, `git pull rust master` to
+update it with the most recent changes.
+
+### Rebasing
+
+You're now ready to start the rebasing process. Check out the branch with your
+changes, and then execute `git rebase master`.
+
+First, a little background: In git, commits are stored as "diffs" which are a
+record of all the changes that a commit made to its parent. When you rebase a
+branch, all the changes in the commits on that branch are reapplied on the
+branch you are rebasing on top of (in this case master). In other words, git
+tries to pretend that the changes you made to the old version of master were
+instead made to the new version of master.
+
+During rebasing, you should expect to encounter at least one "rebase conflict."
+This happens when git's attempt to reapply the changes onto the more recent
+version of master fails because your changes conflicted with other changes that
+have been made since then. You can tell that this happened because you'll see
+lines in the output that look like
+
+```
+CONFLICT (content): Merge conflict in file.rs
+```
+
+When you open these files, you'll see sections of the form
+
+```
+<<<<<<< HEAD
+Original code
+=======
+Your code
+>>>>>>> 8fbf656... Commit fixes 12345
+```
+
+This represents the lines in the file that git could not figure out how to
+rebase. The section between `<<<<<<< HEAD` and `=======` has the code from
+master, while the other side has your version of the code. You'll need to
+decide how to deal with the conflict. You may want to keep your changes,
+keep the changes on master, or combine the two.
+ 
+Generally, resovling the conflict consists of two steps: First, fix the
+particular conflict. Edit the file to make the changes you want and remove the
+`<<<<<<<`, `=======` and `>>>>>>>` lines in the process. Second, check the
+surrounding code. If there was a conflict, its because someone else changed the
+same code you did. That means its likely there are some logical errors lying
+around too!
+
+Once you're all done fixing the conflicts, you need to stage the files that had
+conflicts in them via `git add`. Afterwards, run `git rebase --continue` to let
+git know that you've resolved the conflicts and it should finish the rebase.
+Finally, once the rebase has succeeded, you'll want to update the associated
+branch on your fork with `git push -f`.
```

Comment:
the backticks are weird because github doesn't let me nest them in suggestions
```suggestion
branch on your fork with `git push -f`.

Note that `git push` will not work properly and say something like this:

\```
 ! [rejected]        issue-xxxxx -> issue-xxxxx (non-fast-forward)
error: failed to push some refs to 'https://github.com/username/rust.git'
hint: Updates were rejected because the tip of your current branch is behind
hint: its remote counterpart. Integrate the remote changes (e.g.
hint: 'git pull ...') before pushing again.
hint: See the 'Note about fast-forwards' in 'git push --help' for details.
```

The advice it gives here is incorrect! Because rust has a 'no-merge' policy, running `git pull` will create a merge commit, defeating the point of your rebase. Use `git push -f` instead.
```